pkg/backend/s3/template.go

```diff
@@ -14,16 +14,16 @@
 
 package s3
 
-var backendConfigTmpl = `
-backend "s3" {
-  bucket = "{{ .Bucket }}"
-  key    = "{{ .Key }}"
-  region = "{{ .Region }}"
-  {{- if .AWSCredsPath }}
-  shared_credentials_file = "{{ .AWSCredsPath }}"
-  {{- end }}
-  {{- if .DynamoDBTable }}
-  dynamodb_table = "{{ .DynamoDBTable }}"
-  {{- end }}
-}
-`
+var backendConfigTmpl = `terraform {
```

Comment:
Do we have these other things now? If not, it might be better to refactor when we do so that it's clear how to implement this.

Reply:
I'd say yes, as the experimental PR for Terraform 0.13 is opened (#824) and I'm in the process of adding a CI support for it.

Also, with such implementation, you're changing the behavior (IIUC), that now backend configuration will always be present, regardless whether it was actually configured by the user or not. Currently implementation skips writing the file if it's empty.

On the other hand, I see now that it is permitted to have multiple `terraform {}` blocks. It is only not permitted to have multiple `terraform {}` blocks with `required_providers {}` sub-block. But I'm not sure if we want to have multiple ones.

Reply:
>Also, with such implementation, you're changing the behavior (IIUC), that now backend configuration will always be present, regardless whether it was actually configured by the user or not. Currently implementation skips writing the file if it's empty.

Nope, we're still checking: https://github.com/kinvolk/lokomotive/blob/9824ce7977b83fc74ae726d62fa3507ec01ee00c/pkg/terraform/util.go#L36-L39

We could probably improve the way we render `cluster.tf` so that we collect data from multiple sources (platform, backend etc.) and the render `cluster.tf` "in one go", however doing so exceeds the scope of this PR.

Reply:
I'll undo the `terraform {}` part and look for a good way to deal with the contents of this block.

Reply:
>I'll undo the terraform {} part and look for a good way to deal with the contents of this block.

Thanks, I guess that make sense :+1: 

>I think that pkg/terraform/util.go is a terrible place for writing the backend file 

To be honest, that depends. creating this file is actually the Terraform-related functionality, so it loosely belongs to `terraform` package.

Reply:
I don't think that the `terraform` package should be writing files at all.

Reply:
Well, terraform package could take Terraform code as an input, file is a format used by Terraform UI then :smile:

Reply:
I've undone the `terraform {}` block change. We now render things they way we did until now minus some whitespace problems I've fixed.

As for a better mechanism for rendering the `terraform {}` block and possibly the entire root module - I don't want to do it in this PR as it will quickly grow into yet another big refactoring journey. I'm already touching the root module generation in the platform refactoring PR (which I'm still struggling to get to) so let's see how things look after that.